### PR TITLE
[AIP-44] Migrate run_task_by_local_task_job to Internal API

### LINKED ISSUE #35526: AIP-44 Migrate TaskInstance._run_task_by_local_task_job to Internal API

### PR DESCRIPTION
**Pull request by mhenc**
closes: https://github.com/apache/airflow/issues/35526
related: https://github.com/apache/airflow/issues/35526

### CODE REVIEW

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -2025,6 +2060,7 @@ def check_and_change_state_before_execution(
         ignore_ti_state: bool = False,
         mark_success: bool = False,
         test_mode: bool = False,
+        hostname: str = "",
```
Comment (vincbeck) [inline airflow/models/taskinstance.py:2067]: Missing in docstring

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -3196,19 +3269,30 @@ def schedule_downstream_tasks(self, session: Session = NEW_SESSION, max_tis_per_
                     schedulable_ti.task = task.dag.get_task(schedulable_ti.task_id)
 
             num = dag_run.schedule_tis(schedulable_tis, session=session, max_tis_per_query=max_tis_per_query)
-            self.log.info("%d downstream tasks scheduled from follow-on schedule check", num)
+            ti.log.info("%d downstream tasks scheduled from follow-on schedule check", num)
```
Comment (vincbeck) [inline airflow/models/taskinstance.py]: Would not be better to make it a `classmethod` and use `cls.log`?

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (mhenc) [reply airflow/models/taskinstance.py:2067]: Thanks. Done.

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (mhenc) [reply airflow/models/taskinstance.py]: Good idea

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -2107,47 +2147,80 @@ def check_and_change_state_before_execution(
                 ignore_ti_state=ignore_ti_state,
                 description="requeueable deps",
             )
-            if not self.are_dependencies_met(dep_context=dep_context, session=session, verbose=True):
-                self.state = None
-                self.log.warning(
+            if not ti.are_dependencies_met(dep_context=dep_context, session=session, verbose=True):
+                ti.state = None
+                ti.log.warning(
```
Comment (vincbeck) [inline airflow/models/taskinstance.py]: Same here for log

File: airflow/models/taskinstance.py [hunk 4]
```diff
@@ -2107,47 +2147,80 @@ def check_and_change_state_before_execution(
...
-                session.merge(self)
+                ti.queued_dttm = timezone.utcnow()
+                session.merge(ti)
                 session.commit()
                 return False
 
-        if self.next_kwargs is not None:
-            self.log.info("Resuming after deferral")
+        if ti.next_kwargs is not None:
+            ti.log.info("Resuming after deferral")
```
Comment (vincbeck) [inline airflow/models/taskinstance.py]: Same

File: airflow/models/taskinstance.py [hunk 5]
```diff
@@ -2107,47 +2147,80 @@ def check_and_change_state_before_execution(
...
         else:
-            self.log.info("Starting attempt %s of %s", self.try_number, self.max_tries + 1)
-        self._try_number += 1
+            ti.log.info("Starting attempt %s of %s", ti.try_number, ti.max_tries + 1)
```
Comment (vincbeck) [inline airflow/models/taskinstance.py]: Same

File: airflow/models/taskinstance.py [hunk 6]
```diff
@@ -2107,47 +2147,80 @@ def check_and_change_state_before_execution(
...
+            session.merge(ti).task = task
         session.commit()
 
         # Closing all pooled connections to prevent
         # "max number of connections reached"
         settings.engine.dispose()  # type: ignore
         if verbose:
             if mark_success:
-                self.log.info("Marking success for %s on %s", self.task, self.execution_date)
+                ti.log.info("Marking success for %s on %s", ti.task, ti.execution_date)
```
Comment (vincbeck) [inline airflow/models/taskinstance.py]: Same

File: airflow/models/taskinstance.py [hunk 7]
```diff
@@ -2107,47 +2147,80 @@ def check_and_change_state_before_execution(
...
             else:
-                self.log.info("Executing %s on %s", self.task, self.execution_date)
+                ti.log.info("Executing %s on %s", ti.task, ti.execution_date)
```
Comment (vincbeck) [inline airflow/models/taskinstance.py]: Same

**Review by vincbeck — APPROVED** [review]

**Comment by vincbeck** [comment]
Tests are failing but LGTM

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Yep. Looks good.

**Review by potiuk — APPROVED** [review]

**Pull request merged**